<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Exploring DESeq2 results: Wald test</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="32_wald_files/libs/clipboard/clipboard.min.js"></script>
<script src="32_wald_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="32_wald_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="32_wald_files/libs/quarto-html/popper.min.js"></script>
<script src="32_wald_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="32_wald_files/libs/quarto-html/anchor.min.js"></script>
<link href="32_wald_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="32_wald_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="32_wald_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="32_wald_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="32_wald_files/libs/bootstrap/bootstrap-05cc275de2cb3139844e2d4d6308aaf1.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploring DESeq2 results: Wald test</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>Discuss the steps required to generate a results table for pairwise comparisons (Wald test)</li>
<li>Summarize the different levels of gene filtering</li>
<li>Explain log fold change shrinkage</li>
</ul>
</section>
<section id="exploring-results-wald-test" class="level1">
<h1>Exploring Results (Wald test)</h1>
<p>By default DESeq2 uses the Wald test to identify genes that are differentially expressed between two sample classes. Given the factor(s) used in the design formula, and how many factor levels are present, we can extract results for a number of different comparisons. Here, we will walk through how to obtain results from the <code>dds</code> object and provide some explanations on how to interpret them.</p>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> The Wald test can also be used with <strong>continuous variables</strong>. If the variable of interest provided in the design formula is continuous-valued, then the reported <code>log2FoldChange</code> is per unit of change of that variable.</p>
</blockquote>
<section id="specifying-contrasts" class="level2">
<h2 class="anchored" data-anchor-id="specifying-contrasts">Specifying contrasts</h2>
<p>In our dataset, we have a single variable in our design formula which has three sample classes (levels) so we can make three possible pairwise comparisons:</p>
<ol type="1">
<li>Control vs.&nbsp;Mov10 overexpression</li>
<li>Control vs.&nbsp;Mov10 knockdown</li>
<li>Mov10 knockdown vs.&nbsp;Mov10 overexpression</li>
</ol>
<p><strong>We are really only interested in #1 and #2 from above</strong>. When we intially created our <code>dds</code> object we had provided <code>~ sampletype</code> as our design formula, indicating that <code>sampletype</code> is our main factor of interest.</p>
<p>To indicate which two sample classes we are interested in comparing, we need to specify <strong>contrasts</strong>. The contrasts are used as input to the DESeq2 <code>results()</code> function to extract the desired results.</p>
<blockquote class="blockquote">
<p><strong><em>NOTE:</em> If we run the <code>results()</code> function without specifying <code>contrast</code> or <code>name</code>, it will return the comparison of the last level of the last variable in the design formula over the first level of this variable.</strong> If the order of levels are not specified, they are ordered alphabetically by DESeq2.</p>
</blockquote>
<p>Contrasts can be specified in two different ways (with the first method more commonly used):</p>
<ol type="1">
<li>Contrasts can be supplied as a <strong>character vector with exactly three elements</strong>: the name of the factor (of interest) in the design formula, the name of the two factors levels to compare. The factor level given last is the base level for the comparison. The syntax is given below:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"condition"</span>, <span class="st">"level_to_compare"</span>, <span class="st">"base_level"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(dds, <span class="at">contrast =</span> contrast)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li>Contrasts can be given as a <strong>list of 2 character vectors</strong>: the names of the fold changes for the level of interest, and the names of the fold changes for the base level. These names should match identically to the elements of <code>resultsNames(object)</code>. <em>This method can be useful for combining interaction terms and main effects.</em></li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds) <span class="co"># to see what names to use</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">resultsNames</span>(dds)[<span class="dv">1</span>], <span class="fu">resultsNames</span>(dds)[<span class="dv">2</span>])</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(dds, <span class="at">contrast =</span> contrast)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Alternatively, if you <strong>only had two factor levels you could do nothing</strong> and not worry about specifying contrasts (i.e.&nbsp;<code>results(dds)</code>). In this case, DESeq2 will choose what your base factor level based on alphabetical order of the levels.</p>
<p>To start, we want to evaluate <strong>expression changes between the MOV10 overexpression samples and the control samples</strong>. As such we will use the first method for specifcying contrasts and create a character vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load the libraries</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DEGreport)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tximport)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ashr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read in the object</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>txi <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../../out/object/txi.rds"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../../out/object/metadata.rds"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="do">## Create DESeq2Dataset object</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span>(txi, <span class="at">colData =</span> meta, <span class="at">design =</span> <span class="sc">~</span> sampletype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using just counts from tximport</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 57761 8 
metadata(1): version
assays(1): counts
rownames(57761): ENSG00000000003 ENSG00000000005 ... ENSG00000285978
  ENSG00000285980
rowData names(0):
colnames(8): Irrel_kd_1 Irrel_kd_2 ... Mov10_oe_2 Mov10_oe_3
colData names(1): sampletype</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Run analysis</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check the coefficients based on the results</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Intercept"                                 
[2] "sampletype_MOV10_knockdown_vs_control"     
[3] "sampletype_MOV10_overexpression_vs_control"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Define contrasts for MOV10 overexpression</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>contrast_oe <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"sampletype"</span>, <span class="st">"MOV10_overexpression"</span>, <span class="st">"control"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Contrast can also be build with the <code>makeContrasts</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># build the design object</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>sampletype <span class="ot">&lt;-</span> meta<span class="sc">$</span>sampletype</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>sampletype)</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># print the model.matrix</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (Intercept) sampletypeMOV10_knockdown sampletypeMOV10_overexpression
1           1                         0                              0
2           1                         0                              0
3           1                         0                              0
4           1                         1                              0
5           1                         1                              0
6           1                         0                              1
7           1                         0                              1
8           1                         0                              1
attr(,"assign")
[1] 0 1 1
attr(,"contrasts")
attr(,"contrasts")$sampletype
[1] "contr.treatment"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="at">MOV10KO_vs_CTRL =</span> sampletypeMOV10_knockdown,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">MOV10OE_vs_CTRL =</span> sampletypeMOV10_overexpression,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">MOV10KO_vs_MOV10OE =</span> sampletypeMOV10_knockdown <span class="sc">-</span> sampletypeMOV10_overexpression,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> design)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in makeContrasts(MOV10KO_vs_CTRL = sampletypeMOV10_knockdown,
MOV10OE_vs_CTRL = sampletypeMOV10_overexpression, : Renaming (Intercept) to
Intercept</code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>contrast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                Contrasts
Levels                           MOV10KO_vs_CTRL MOV10OE_vs_CTRL
  Intercept                                    0               0
  sampletypeMOV10_knockdown                    1               0
  sampletypeMOV10_overexpression               0               1
                                Contrasts
Levels                           MOV10KO_vs_MOV10OE
  Intercept                                       0
  sampletypeMOV10_knockdown                       1
  sampletypeMOV10_overexpression                 -1</code></pre>
</div>
</div>
<blockquote class="blockquote">
<h3 id="does-it-matter-what-i-choose-to-be-my-base-level" class="anchored">Does it matter what I choose to be my base level?</h3>
<p>Yes, it does matter. <strong>Deciding what level is the base level will determine how to interpret the fold change that is reported.</strong> So for example, if we observe a log2 fold change of -2 this would mean the gene expression is lower in factor level of interest relative to the base level. Thus, if leaving it up to DESeq2 to decide on the contrasts be sure to check that the alphabetical order coincides with the fold change direction you are anticipating.</p>
</blockquote>
</section>
<section id="the-results-table" class="level2">
<h2 class="anchored" data-anchor-id="the-results-table">The results table</h2>
<p>Now that we have our contrast created, we can use it as input to the <code>results()</code> function. Let’s take a quick look at the help manual for the function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>?results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You will see we have the option to provide a wide array of arguments and tweak things from the defaults as needed. As we go through the lesson we will keep coming back to the help documentation to discuss some arguments that are good to know about.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract results for MOV10 overexpression vs control</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast_oe, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>res_tableOE</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): sampletype MOV10_overexpression vs control 
Wald test p-value: sampletype MOV10 overexpression vs control 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE       stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835      -0.438245 0.0774607 -5.6576468 1.53463e-08
ENSG00000000005    26.2489       0.029208 0.4411289  0.0662119 9.47209e-01
ENSG00000000419  1478.2512       0.383635 0.1137610  3.3722909 7.45457e-04
ENSG00000000457   518.4220       0.228971 0.1023312  2.2375448 2.52508e-02
ENSG00000000460  1159.7761      -0.269138 0.0814992 -3.3023396 9.58819e-04
...                    ...            ...       ...        ...         ...
ENSG00000285889    1.82171       -4.68144 3.9266061   -1.19224  0.23316908
ENSG00000285950    7.58089       -1.01978 1.0715579   -0.95168  0.34125937
ENSG00000285976 4676.24904        0.19364 0.0656673    2.94881  0.00319004
ENSG00000285978    2.25697        4.13612 2.0706198    1.99753  0.04576780
ENSG00000285980    0.00000             NA        NA         NA          NA
                       padj
                  &lt;numeric&gt;
ENSG00000000003 4.25096e-07
ENSG00000000005 9.72687e-01
ENSG00000000419 4.67394e-03
ENSG00000000457 8.02341e-02
ENSG00000000460 5.76079e-03
...                     ...
ENSG00000285889          NA
ENSG00000285950          NA
ENSG00000285976   0.0157502
ENSG00000285978          NA
ENSG00000285980          NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># try to extract the same table from the makeContrasts construct</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>res_tableOE_alt <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast[,<span class="st">"MOV10OE_vs_CTRL"</span>], <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># confirm the two tables are the same</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>res_tableOE_alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): 0,0,+1 
Wald test p-value: 0,0,+1 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE       stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835      -0.438245 0.0774607 -5.6576468 1.53463e-08
ENSG00000000005    26.2489       0.029208 0.4411289  0.0662119 9.47209e-01
ENSG00000000419  1478.2512       0.383635 0.1137610  3.3722909 7.45457e-04
ENSG00000000457   518.4220       0.228971 0.1023312  2.2375448 2.52508e-02
ENSG00000000460  1159.7761      -0.269138 0.0814992 -3.3023396 9.58819e-04
...                    ...            ...       ...        ...         ...
ENSG00000285889    1.82171       -4.68144 3.9266061   -1.19224  0.23316908
ENSG00000285950    7.58089       -1.01978 1.0715579   -0.95168  0.34125937
ENSG00000285976 4676.24904        0.19364 0.0656673    2.94881  0.00319004
ENSG00000285978    2.25697        4.13612 2.0706198    1.99753  0.04576780
ENSG00000285980    0.00000             NA        NA         NA          NA
                       padj
                  &lt;numeric&gt;
ENSG00000000003 4.25096e-07
ENSG00000000005 9.72687e-01
ENSG00000000419 4.67394e-03
ENSG00000000457 8.02341e-02
ENSG00000000460 5.76079e-03
...                     ...
ENSG00000285889          NA
ENSG00000285950          NA
ENSG00000285976   0.0157502
ENSG00000285978          NA
ENSG00000285980          NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>res_tableOE</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): sampletype MOV10_overexpression vs control 
Wald test p-value: sampletype MOV10 overexpression vs control 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE       stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835      -0.438245 0.0774607 -5.6576468 1.53463e-08
ENSG00000000005    26.2489       0.029208 0.4411289  0.0662119 9.47209e-01
ENSG00000000419  1478.2512       0.383635 0.1137610  3.3722909 7.45457e-04
ENSG00000000457   518.4220       0.228971 0.1023312  2.2375448 2.52508e-02
ENSG00000000460  1159.7761      -0.269138 0.0814992 -3.3023396 9.58819e-04
...                    ...            ...       ...        ...         ...
ENSG00000285889    1.82171       -4.68144 3.9266061   -1.19224  0.23316908
ENSG00000285950    7.58089       -1.01978 1.0715579   -0.95168  0.34125937
ENSG00000285976 4676.24904        0.19364 0.0656673    2.94881  0.00319004
ENSG00000285978    2.25697        4.13612 2.0706198    1.99753  0.04576780
ENSG00000285980    0.00000             NA        NA         NA          NA
                       padj
                  &lt;numeric&gt;
ENSG00000000003 4.25096e-07
ENSG00000000005 9.72687e-01
ENSG00000000419 4.67394e-03
ENSG00000000457 8.02341e-02
ENSG00000000460 5.76079e-03
...                     ...
ENSG00000285889          NA
ENSG00000285950          NA
ENSG00000285976   0.0157502
ENSG00000285978          NA
ENSG00000285980          NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res_tableOE_alt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 2011, 5.2%
LFC &lt; 0 (down)     : 2797, 7.2%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res_tableOE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 2011, 5.2%
LFC &lt; 0 (down)     : 2797, 7.2%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> For our analysis, in addition to the <code>contrast</code> argument we will also provide a value of 0.05 for the <code>alpha</code> argument. We will describe this in more detail when we talk about <a href="">gene-level filtering</a>.</p>
</blockquote>
<p>The results table that is returned to us is <strong>a <code>DESeqResults</code> object</strong>, which is a simple subclass of DataFrame. In many ways it can be treated like a dataframe (i.e when accessing/subsetting data), however it is important to recognize that there are differences for downstream steps like visualization.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check what type of object is returned</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(res_tableOE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DESeqResults"
attr(,"package")
[1] "DESeq2"</code></pre>
</div>
</div>
<p>Now let’s take a look at <strong>what information is stored</strong> in the results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What is stored in results?</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    baseMean log2FoldChange      lfcSE        stat       pvalue
ENSG00000000003 3525.8834779    -0.43824542 0.07746072 -5.65764683 1.534625e-08
ENSG00000000005   26.2489043     0.02920799 0.44112891  0.06621191 9.472091e-01
ENSG00000000419 1478.2512405     0.38363504 0.11376096  3.37229086 7.454568e-04
ENSG00000000457  518.4220238     0.22897058 0.10233117  2.23754477 2.525076e-02
ENSG00000000460 1159.7761365    -0.26913820 0.08149925 -3.30233963 9.588190e-04
ENSG00000000938    0.4021325    -1.79625188 4.02392248 -0.44639326 6.553132e-01
                        padj
ENSG00000000003 4.250960e-07
ENSG00000000005 9.726866e-01
ENSG00000000419 4.673942e-03
ENSG00000000457 8.023408e-02
ENSG00000000460 5.760786e-03
ENSG00000000938           NA</code></pre>
</div>
</div>
<p>We have six columns of information reported for each gene (row). We can use the <code>mcols()</code> function to extract information on what the values stored in each column represent:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get information on each column in results</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mcols</span>(res_tableOE, <span class="at">use.names=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 2 columns
                       type            description
                &lt;character&gt;            &lt;character&gt;
baseMean       intermediate mean of normalized c..
log2FoldChange      results log2 fold change (ML..
lfcSE               results standard error: samp..
stat                results Wald statistic: samp..
pvalue              results Wald test p-value: s..
padj                results   BH adjusted p-values</code></pre>
</div>
</div>
<ul>
<li><code>baseMean</code>: mean of normalized counts for all samples</li>
<li><code>log2FoldChange</code>: log2 fold change</li>
<li><code>lfcSE</code>: standard error</li>
<li><code>stat</code>: Wald statistic</li>
<li><code>pvalue</code>: Wald test p-value</li>
<li><code>padj</code>: BH adjusted p-values</li>
</ul>
</section>
<section id="p-values" class="level2">
<h2 class="anchored" data-anchor-id="p-values">P-values</h2>
<p>The p-value is a probability value used to determine whether there is evidence to reject the null hypothesis. <strong>A smaller p-value means that there is stronger evidence in favor of the alternative hypothesis</strong>. However, because we are performing a test for each individual gene we need to correct these p-values for multiple testing.</p>
<p><strong>The <code>padj</code> column</strong> in the results table represents the p-value adjusted for multiple testing, and is the most important column of the results. Typically, a threshold such as <code>padj</code> &lt; 0.05 is a good starting point for identifying significant genes. The default method for <strong>multiple test correction</strong> in DESeq2 is an implementation of the Benjamini-Hochberg false discovery rate (FDR). There are other corrections methods available and can be changed by adding the <code>pAdjustMethod</code> argument to the <code>results()</code> function.</p>
<section id="gene-level-filtering" class="level3">
<h3 class="anchored" data-anchor-id="gene-level-filtering">Gene-level filtering</h3>
<p>Let’s take a closer look at our results table. As we scroll through it, you will notice that for <strong>selected genes there are NA values in the <code>pvalue</code> and <code>padj</code> columns</strong>. What does this mean?</p>
<div id="fig-1" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/gene_filtering.png" class="img-fluid figure-img" width="800"></p>
<figcaption>1</figcaption>
</figure>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1
</figcaption>
</figure>
</div>
<p>The missing values represent genes that have undergone filtering as part of the <code>DESeq()</code> function. Prior to differential expression analysis it is <strong>beneficial to omit genes that have little or no chance of being detected as differentially expressed.</strong> This will increase the power to detect differentially expressed genes. DESeq2 does not physically remove any genes from the original counts matrix, and so all genes will be present in your results table. The genes omitted by DESeq2 meet one of the <strong>three filtering criteria outlined below</strong>:</p>
<p><strong>1. Genes with zero counts in all samples</strong></p>
<p>If within a row, all samples have zero counts there is no expression information and therefore these genes are not tested.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter genes by zero expression</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>res_tableOE[<span class="fu">which</span>(res_tableOE<span class="sc">$</span>baseMean <span class="sc">==</span> <span class="dv">0</span>),] <span class="sc">%&gt;%</span> </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                baseMean log2FoldChange lfcSE stat pvalue padj
ENSG00000005381        0             NA    NA   NA     NA   NA
ENSG00000008196        0             NA    NA   NA     NA   NA
ENSG00000008197        0             NA    NA   NA     NA   NA
ENSG00000008438        0             NA    NA   NA     NA   NA
ENSG00000011083        0             NA    NA   NA     NA   NA
ENSG00000011600        0             NA    NA   NA     NA   NA</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>The baseMean column for these genes will be zero, and the log2 fold change estimates, p-value and adjusted p-value will all be set to NA.</strong></p>
</blockquote>
<p><strong>2. Genes with an extreme count outlier</strong></p>
<p>The <code>DESeq()</code> function calculates, for every gene and for every sample, a diagnostic test for outliers called Cook’s distance. Cook’s distance is a measure of how much a single sample is influencing the fitted coefficients for a gene, and a large value of Cook’s distance is intended to indicate an outlier count. Genes which contain a Cook’s distance above a threshold are flagged, however at least 3 replicates are required for flagging, as it is difficult to judge which sample might be an outlier with only 2 replicates. We can turn off this filtering by using the <code>cooksCutoff</code> argument in the <code>results()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter genes that have an extreme outlier</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">is.na</span>(pvalue) <span class="sc">&amp;</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>           <span class="fu">is.na</span>(padj) <span class="sc">&amp;</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>           baseMean <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 baseMean log2FoldChange    lfcSE       stat pvalue padj
ENSG00000001626 123.83122     -0.3561765 3.438228 -0.1035930     NA   NA
ENSG00000008226  67.82651      5.8902223 1.973889  2.9840702     NA   NA
ENSG00000058866 116.38549      0.8519290 2.133619  0.3992882     NA   NA
ENSG00000060709 108.76629      4.5901999 1.716474  2.6742036     NA   NA
ENSG00000100346 121.65513      6.3764242 1.803765  3.5350630     NA   NA
ENSG00000100427  37.51051     -2.1283061 1.491536 -1.4269224     NA   NA</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>If a gene contains a sample with an extreme count outlier then the p-value and adjusted p-value will be set to NA.</strong></p>
</blockquote>
<p><strong>3. Genes with a low mean normalized counts</strong></p>
<p>DESeq2 defines a low mean threshold, that is empirically determined from your data, in which the fraction of significant genes can be increased by reducing the number of genes that are considered for multiple testing. This is based on the notion that genes with very low counts are not likely to see significant differences typically due to high dispersion.</p>
<div id="fig-2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/indep_filt_scatterplus.png" class="img-fluid figure-img" width="800"></p>
<figcaption>2</figcaption>
</figure>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2
</figcaption>
</figure>
</div>
<p><em>Image courtesy of <a href="https://www.slideshare.net/joachimjacob/5rna-seqpart5detecting-differentialexpression">slideshare presentation</a> from Joachim Jacob, 2014.</em></p>
<p>At a user-specified value (<code>alpha = 0.05</code>), DESeq2 evaluates the change in the number of significant genes as it filters out increasingly bigger portions of genes based on their mean counts, as shown in the figure above. The point at which the number of significant genes reaches a peak is the low mean threshold that is used to filter genes that undergo multiple testing. There is also an argument to turn off the filtering off by setting <code>independentFiltering = F</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter genes below the low mean threshold</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(pvalue) <span class="sc">&amp;</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>           <span class="fu">is.na</span>(padj) <span class="sc">&amp;</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>           baseMean <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">%&gt;%</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  baseMean log2FoldChange     lfcSE       stat     pvalue padj
ENSG00000000938  0.4021325     -1.7962519 4.0239225 -0.4463933 0.65531320   NA
ENSG00000000971 11.3963043     -0.6130577 0.7765893 -0.7894233 0.42986462   NA
ENSG00000002079  4.3709939     -0.4818745 1.5967743 -0.3017800 0.76281981   NA
ENSG00000002726  1.7387962     -0.3834633 1.9626055 -0.1953848 0.84509173   NA
ENSG00000002745  7.2065290      0.3893226 0.7689257  0.5063202 0.61263186   NA
ENSG00000002933 10.6169395     -1.4313952 0.6449315 -2.2194529 0.02645593   NA</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>If a gene is filtered by independent filtering, then only the adjusted p-value will be set to NA.</strong></p>
</blockquote>
<blockquote class="blockquote">
<p><em><strong>NOTE:</strong> DESeq2 will perform the filtering outlined above by default; however other DE tools, such as EdgeR will not. Filtering is a necessary step, even if you are using limma-voom and/or edgeR’s quasi-likelihood methods. Be sure to follow pre-filtering steps when using other tools, as outlined in their user guides found on Bioconductor as they generally perform much better.</em></p>
</blockquote>
</section>
</section>
<section id="fold-change" class="level2">
<h2 class="anchored" data-anchor-id="fold-change">Fold change</h2>
<p>Another important column in the results table, is the <code>log2FoldChange</code>. With large significant gene lists it can be hard to extract meaningful biological relevance. To help increase stringency, one can also <strong>add a fold change threshold</strong>. Keep in mind when setting that value that we are working with log2 fold changes, so a cutoff of <code>log2FoldChange</code> &lt; 1 would translate to an actual fold change of 2.</p>
<blockquote class="blockquote">
<h3 id="an-alternative-approach-to-add-the-fold-change-threshold" class="anchored">An alternative approach to add the fold change threshold:</h3>
<p>The <code>results()</code> function has an option to add a fold change threshold using the <code>lfcThrehsold</code> argument. This method is more statistically motivated, and is recommended when you want a more confident set of genes based on a certain fold-change. It actually performs a statistical test against the desired threshold, by performing a two-tailed test for log2 fold changes greater than the absolute value specified. The user can change the alternative hypothesis using <code>altHypothesis</code> and perform two one-tailed tests as well. <strong>This is a more conservative approach, so expect to retrieve a much smaller set of genes!</strong></p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see the two different summaries using the FC threshold or not</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast_oe, <span class="at">alpha =</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span> <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 2011, 5.2%
LFC &lt; 0 (down)     : 2797, 7.2%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast_oe, <span class="at">alpha =</span> <span class="fl">0.05</span>,<span class="at">lfcThreshold =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 1.00 (up)    : 16, 0.041%
LFC &lt; -1.00 (down) : 7, 0.018%
outliers [1]       : 28, 0.072%
low counts [2]     : 16905, 43%
(mean count &lt; 5)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># confirm that this is not the same as filtering over the original table</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="sc">%&gt;%</span> <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 2011, 5.2%
LFC &lt; 0 (down)     : 2797, 7.2%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter the table without the threshold</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>res_tableOE <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">test =</span> <span class="fu">case_when</span>(log2FoldChange<span class="sc">&gt;</span><span class="dv">1</span> <span class="sc">&amp;</span> padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"sig-up"</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>                          log2FoldChange<span class="sc">&lt;</span>(<span class="sc">-</span><span class="dv">1</span>) <span class="sc">&amp;</span> padj <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">~</span> <span class="st">"sig-down"</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>                          T <span class="sc">~</span> <span class="st">"non-sig"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(test) <span class="sc">%&gt;%</span></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">n =</span> <span class="fu">n</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  test         n
  &lt;chr&gt;    &lt;int&gt;
1 non-sig  57293
2 sig-down   159
3 sig-up     309</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># see the output table after using the FCC threhold</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast_oe, <span class="at">alpha =</span> <span class="fl">0.05</span>,<span class="at">lfcThreshold =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): sampletype MOV10_overexpression vs control 
Wald test p-value: sampletype MOV10 overexpression vs control 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE       stat    pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
ENSG00000000003  3525.8835      -0.438245 0.0774607 -5.6576468  1.000000
ENSG00000000005    26.2489       0.029208 0.4411289  0.0662119  0.995942
ENSG00000000419  1478.2512       0.383635 0.1137610  3.3722909  1.000000
ENSG00000000457   518.4220       0.228971 0.1023312  2.2375448  1.000000
ENSG00000000460  1159.7761      -0.269138 0.0814992 -3.3023396  1.000000
...                    ...            ...       ...        ...       ...
ENSG00000285889    1.82171       -4.68144 3.9266061   -1.19224 0.2481961
ENSG00000285950    7.58089       -1.01978 1.0715579   -0.95168 0.5223579
ENSG00000285976 4676.24904        0.19364 0.0656673    2.94881 1.0000000
ENSG00000285978    2.25697        4.13612 2.0706198    1.99753 0.0714996
ENSG00000285980    0.00000             NA        NA         NA        NA
                     padj
                &lt;numeric&gt;
ENSG00000000003         1
ENSG00000000005         1
ENSG00000000419         1
ENSG00000000457         1
ENSG00000000460         1
...                   ...
ENSG00000285889        NA
ENSG00000285950         1
ENSG00000285976         1
ENSG00000285978        NA
ENSG00000285980        NA</code></pre>
</div>
</div>
<p>The fold changes reported in the results table are calculated by:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">log2</span> (normalized_counts_group1 <span class="sc">/</span> normalized_counts_group2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The problem is, these fold change estimates are not entirely accurate as they do not account for the large dispersion we observe with low read counts. To address this, the <strong>log2 fold changes need to be adjusted</strong>.</p>
<section id="more-accurate-lfc-estimates" class="level3">
<h3 class="anchored" data-anchor-id="more-accurate-lfc-estimates">More accurate LFC estimates</h3>
<p>To generate more accurate log2 foldchange (LFC) estimates, DESeq2 allows for the <strong>shrinkage of the LFC estimates toward zero</strong> when the information for a gene is low, which could include:</p>
<ul>
<li>Low counts</li>
<li>High dispersion values</li>
</ul>
<p>LFC shrinkage uses <strong>information from all genes</strong> to generate more accurate estimates. Specifically, the distribution of LFC estimates for all genes is used (as a prior) to shrink the LFC estimates of genes with little information or high dispersion toward more likely (lower) LFC estimates.</p>
<div id="fig-2" class="quarto-float quarto-figure quarto-figure-center anchored" data-fig-align="center">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/deseq2_shrunken_lfc.png" class="img-fluid figure-img" width="500"></p>
<figcaption>2</figcaption>
</figure>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3
</figcaption>
</figure>
</div>
<p><em>Illustration taken from the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2 paper</a>.</em></p>
<p>In the figure above, we have an example using two genes green gene and purple gene. For each gene the expression values are plotted for each sample in the two different mouse strains (C57BL/6J and DBA/2J). Both genes have the same mean values for the two sample groups, but the green gene has little variation within group while the purple gene has high levels of variation. For the green gene with low within group variation, the <strong>unshrunken LFC estimate</strong> (vertex of the green <strong>solid line</strong>) is very similar to the shrunken LFC estimate (vertex of the green dotted line). However, LFC estimates for the purple gene are quite different due to the high dispersion. So even though two genes can have similar normalized count values, they can have differing degrees of LFC shrinkage. Notice the <strong>LFC estimates are shrunken toward the prior (black solid line)</strong>.</p>
<p><strong>Shrinking the log2 fold changes will not change the total number of genes that are identified as significantly differentially expressed.</strong> The shrinkage of fold change is to help with downstream assessment of results. For example, if you wanted to subset your significant genes based on fold change for further evaluation, you may want to use shruken values. Additionally, for functional analysis tools such as GSEA which require fold change values as input you would want to provide shrunken values.</p>
<p>To generate the shrunken log2 fold change estimates, you have to run an additional step on your results object (that we will create below) with the function <code>lfcShrink()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Save the unshrunken results to compare</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="co"># res_tableOE_unshrunken &lt;- res_tableOE</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Intercept"                                 
[2] "sampletype_MOV10_knockdown_vs_control"     
[3] "sampletype_MOV10_overexpression_vs_control"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply fold change shrinkage</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>res_tableOE_shr <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef=</span><span class="st">"sampletype_MOV10_overexpression_vs_control"</span>, <span class="at">type=</span><span class="st">"apeglm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895</code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with apeglm I have problem in setting the shrinkage for coefficinets that are nore available in teh resultsNames(). for esample change of reference. the solution is to change the type to ashr</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>res_tableOE_shrAlt <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds,</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">res =</span> res_tableOE,</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>                                <span class="at">type=</span><span class="st">"ashr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'ashr' for LFC shrinkage. If used in published research, please cite:
    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.
    https://doi.org/10.1093/biostatistics/kxw041</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this is more felxible to allow shrinkage of any comparison</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>contrast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                Contrasts
Levels                           MOV10KO_vs_CTRL MOV10OE_vs_CTRL
  Intercept                                    0               0
  sampletypeMOV10_knockdown                    1               0
  sampletypeMOV10_overexpression               0               1
                                Contrasts
Levels                           MOV10KO_vs_MOV10OE
  Intercept                                       0
  sampletypeMOV10_knockdown                       1
  sampletypeMOV10_overexpression                 -1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>res_tableKO_Alt <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast[,<span class="st">"MOV10KO_vs_CTRL"</span>], <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>res_tableKO_shrAlt <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds,</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">res =</span> res_tableKO_Alt,</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">type=</span><span class="st">"ashr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'ashr' for LFC shrinkage. If used in published research, please cite:
    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.
    https://doi.org/10.1093/biostatistics/kxw041</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>res_tableKO_Alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): 0,+1,0 
Wald test p-value: 0,+1,0 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE      stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835       0.019599 0.0857587  0.228537 0.819228863
ENSG00000000005    26.2489      -0.561368 0.5005704 -1.121456 0.262093841
ENSG00000000419  1478.2512       0.200415 0.1267970  1.580593 0.113971138
ENSG00000000457   518.4220       0.421334 0.1113780  3.782920 0.000154999
ENSG00000000460  1159.7761       0.165758 0.0887861  1.866938 0.061910196
...                    ...            ...       ...       ...         ...
ENSG00000285889    1.82171     -4.9988725 4.4250065 -1.129687   0.2586081
ENSG00000285950    7.58089      0.1244283 1.1406775  0.109083   0.9131368
ENSG00000285976 4676.24904     -0.0641105 0.0733078 -0.874539   0.3818246
ENSG00000285978    2.25697      4.3059932 2.2301751  1.930787   0.0535094
ENSG00000285980    0.00000             NA        NA        NA          NA
                      padj
                 &lt;numeric&gt;
ENSG00000000003 0.92347051
ENSG00000000005 0.51683045
ENSG00000000419 0.31379168
ENSG00000000457 0.00243228
ENSG00000000460 0.21169526
...                    ...
ENSG00000285889         NA
ENSG00000285950         NA
ENSG00000285976   0.635781
ENSG00000285978         NA
ENSG00000285980         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>res_tableKO_shrAlt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MMSE): 0,+1,0 
Wald test p-value: 0,+1,0 
DataFrame with 57761 rows and 5 columns
                  baseMean log2FoldChange     lfcSE      pvalue       padj
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt;
ENSG00000000003  3525.8835      0.0155199 0.0763274 0.819228863 0.92347051
ENSG00000000005    26.2489     -0.0940578 0.2200574 0.262093841 0.51683045
ENSG00000000419  1478.2512      0.1333804 0.1065258 0.113971138 0.31379168
ENSG00000000457   518.4220      0.3369935 0.1124607 0.000154999 0.00243228
ENSG00000000460  1159.7761      0.1312291 0.0801808 0.061910196 0.21169526
...                    ...            ...       ...         ...        ...
ENSG00000285889    1.82171    -0.01814890 0.2707641   0.2586081         NA
ENSG00000285950    7.58089     0.00548265 0.2395407   0.9131368         NA
ENSG00000285976 4676.24904    -0.05374190 0.0672231   0.3818246   0.635781
ENSG00000285978    2.25697     0.06162803 0.2925929   0.0535094         NA
ENSG00000285980    0.00000     0.00000000 0.2726062          NA         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>res_tableKOOE_Alt <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast[,<span class="st">"MOV10KO_vs_MOV10OE"</span>], <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>res_tableKOOE_shrAlt <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">res =</span> res_tableKOOE_Alt,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">type=</span><span class="st">"ashr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'ashr' for LFC shrinkage. If used in published research, please cite:
    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.
    https://doi.org/10.1093/biostatistics/kxw041</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>res_tableKOOE_Alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): 0,+1,-1 
Wald test p-value: 0,+1,-1 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE       stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835       0.457844 0.0861268    5.31593 1.06112e-07
ENSG00000000005    26.2489      -0.590576 0.4992358   -1.18296 2.36825e-01
ENSG00000000419  1478.2512      -0.183221 0.1262197   -1.45160 1.46613e-01
ENSG00000000457   518.4220       0.192363 0.1100786    1.74751 8.05492e-02
ENSG00000000460  1159.7761       0.434896 0.0893840    4.86549 1.14176e-06
...                    ...            ...       ...        ...         ...
ENSG00000285889    1.82171       0.000000 4.5621529  0.0000000 1.000000000
ENSG00000285950    7.58089       1.144208 1.1748720  0.9739005 0.330105946
ENSG00000285976 4676.24904      -0.257751 0.0731437 -3.5238930 0.000425256
ENSG00000285978    2.25697       0.169871 1.9487517  0.0871692 0.930537020
ENSG00000285980    0.00000             NA        NA         NA          NA
                       padj
                  &lt;numeric&gt;
ENSG00000000003 1.91949e-06
ENSG00000000005 4.02931e-01
ENSG00000000419 2.85485e-01
ENSG00000000457 1.83312e-01
ENSG00000000460 1.58712e-05
...                     ...
ENSG00000285889          NA
ENSG00000285950          NA
ENSG00000285976  0.00261137
ENSG00000285978          NA
ENSG00000285980          NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>res_tableKOOE_shrAlt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MMSE): 0,+1,-1 
Wald test p-value: 0,+1,-1 
DataFrame with 57761 rows and 5 columns
                  baseMean log2FoldChange     lfcSE      pvalue        padj
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835       0.426671 0.0836899 1.06112e-07 1.91949e-06
ENSG00000000005    26.2489      -0.182854 0.2876357 2.36825e-01 4.02931e-01
ENSG00000000419  1478.2512      -0.157121 0.1170263 1.46613e-01 2.85485e-01
ENSG00000000457   518.4220       0.170754 0.1038243 8.05492e-02 1.83312e-01
ENSG00000000460  1159.7761       0.402902 0.0865484 1.14176e-06 1.58712e-05
...                    ...            ...       ...         ...         ...
ENSG00000285889    1.82171     0.00000000 0.4066965 1.000000000          NA
ENSG00000285950    7.58089     0.10256452 0.3664672 0.330105946          NA
ENSG00000285976 4676.24904    -0.24416072 0.0712637 0.000425256  0.00261137
ENSG00000285978    2.25697     0.00632641 0.3761716 0.930537020          NA
ENSG00000285980    0.00000     0.00000000 0.4519697          NA          NA</code></pre>
</div>
</div>
<p>Depending on the version of DESeq2 you are using the default <strong>method for shrinkage estimation</strong> will differ. The defaults can be changed by adding the argument <code>type</code> in the <code>lfcShrink()</code> function as we have above. For most recent versions of DESeq2, <code>type="normal"</code> is the default and was the only method in earlier versions. It has been shown that in most situations there are alternative methods that have <a href="https://bioconductor.org/packages/devel/bioc/vignettes/apeglm/inst/doc/apeglm.html">less bias than the ’normal` method</a>, and therefore <strong>we chose to use apeglm</strong>.</p>
<p><em>Note that the <code>stat</code> column is no longer present in the results table.</em> You can compare by inspecting the output of <code>res_tableOE</code> and <code>res_tableOE_unshrunken</code> in the R console. If you run <code>type="apeglm"</code> or <code>type="ashr"</code> you do not get a stat column because these don’t really make sense in the context of the posterior quantities that are returned.</p>
<blockquote class="blockquote">
<p>For more information on shrinkage, the DESeq2 vignette has an <a href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#extended-section-on-shrinkage-estimators">Extended section on shrinkage estimators</a> that is quite useful.</p>
</blockquote>
<blockquote class="blockquote">
<h4 id="contrast-vs-coef" class="anchored"><code>contrast</code> vs <code>coef</code></h4>
<p>When using the alternative methods, rather than using the <code>contrast</code> argument you will be required to specify <code>coef</code>. Using contrast forms an expanded model matrix, treating all factor levels equally, and averages over all distances between all pairs of factor levels to estimate the prior. Using coef, means looking only at that column of the model matrix (so usually that would be one level against the reference level) and estimates the prior for that coefficient from the distribution of those MLE of coefficients. When using coef, the shrinkage depends on which level is chosen as reference.</p>
</blockquote>
<blockquote class="blockquote">
<h4 id="how-do-i-know-what-to-value-to-provide-to-the-coef-argument" class="anchored">How do I know what to value to provide to the <code>coef</code> argument?</h4>
<p>The value you provide here needs to match identically to what is stored in the column header of the coefficients table. To see what values you have to work with you can use <code>resultsNames(dds)</code>.</p>
</blockquote>
</section>
</section>
<section id="ma-plot" class="level2">
<h2 class="anchored" data-anchor-id="ma-plot">MA plot</h2>
<p>A plot that can be useful to exploring our results is the MA plot. The MA plot shows the <strong>mean of the normalized counts versus the log2 foldchanges for all genes tested</strong>. The genes that are significantly DE are colored to be easily identified. This is also a great way to illustrate the effect of LFC shrinkage. The DESeq2 package offers a simple function to generate an MA plot.</p>
<p><strong>Let’s start with the unshrunken results:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MA plot using unshrunken fold changes</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>DESeq2<span class="sc">::</span><span class="fu">plotMA</span>(res_tableOE, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>,<span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="32_wald_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>And now the shrunken results:</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MA plot using shrunken fold changes</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>DESeq2<span class="sc">::</span><span class="fu">plotMA</span>(res_tableOE_shr, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">2</span>,<span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="32_wald_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>On the left you have the unshrunken fold change values plotted and you can see the abundance of scatter for the the lowly expressed genes. That is, many of the low expressors exhibit very high fold changes. After shrinkage, we see the fold changes are much smaller estimates.</p>
<p>In addition to the comparison described above, this plot allows us to evaluate the magnitude of fold changes and how they are distributed relative to mean expression. Generally, we would expect to see significant genes across the full range of expression levels.</p>
<hr>
<p><strong>Excercise</strong></p>
<p><strong>MOV10 Differential Expression Analysis: Control versus Knockdown</strong></p>
<p>Now that we have results for the overexpression results, do the same for the <strong>Control vs.&nbsp;Knockdown samples</strong>.</p>
<ol type="1">
<li>Create a contrast vector called <code>contrast_kd</code>.</li>
<li>Use contrast vector in the <code>results()</code> to extract a results table and store that to a variable called <code>res_tableKD</code>.</li>
<li>Shrink the LFC estimates using <code>lfcShrink()</code> and assign it back to <code>res_tableKD</code>.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Create a contrast vector called contrast_kd.</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>contrast_kd <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"sampletype"</span>, <span class="st">"MOV10_knockdown"</span>, <span class="st">"control"</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Use contrast vector in the results() to extract a results table and store that to a variable called res_tableKD.</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>res_tableKD <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast_kd, <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Shrink the LFC estimates using lfcShrink() and assign it back to res_tableKD.</span></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>res_tableKD_shr <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef=</span><span class="st">"sampletype_MOV10_knockdown_vs_control"</span>, <span class="at">type=</span><span class="st">"apeglm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'apeglm' for LFC shrinkage. If used in published research, please cite:
    Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for
    sequence count data: removing the noise and preserving large differences.
    Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895</code></pre>
</div>
</div>
<p>Alternatively</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># build the design object</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>sampletype <span class="ot">&lt;-</span> meta<span class="sc">$</span>sampletype</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span>sampletype)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="co"># print the model.matrix</span></span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (Intercept) sampletypeMOV10_knockdown sampletypeMOV10_overexpression
1           1                         0                              0
2           1                         0                              0
3           1                         0                              0
4           1                         1                              0
5           1                         1                              0
6           1                         0                              1
7           1                         0                              1
8           1                         0                              1
attr(,"assign")
[1] 0 1 1
attr(,"contrasts")
attr(,"contrasts")$sampletype
[1] "contr.treatment"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>contrast <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(<span class="at">MOV10KO_vs_CTRL =</span> sampletypeMOV10_knockdown,</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">MOV10OE_vs_CTRL =</span> sampletypeMOV10_overexpression,</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">MOV10KO_vs_MOV10OE =</span> sampletypeMOV10_knockdown <span class="sc">-</span> sampletypeMOV10_overexpression,</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">levels =</span> design)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in makeContrasts(MOV10KO_vs_CTRL = sampletypeMOV10_knockdown,
MOV10OE_vs_CTRL = sampletypeMOV10_overexpression, : Renaming (Intercept) to
Intercept</code></pre>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>contrast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                Contrasts
Levels                           MOV10KO_vs_CTRL MOV10OE_vs_CTRL
  Intercept                                    0               0
  sampletypeMOV10_knockdown                    1               0
  sampletypeMOV10_overexpression               0               1
                                Contrasts
Levels                           MOV10KO_vs_MOV10OE
  Intercept                                       0
  sampletypeMOV10_knockdown                       1
  sampletypeMOV10_overexpression                 -1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>res_tableKO_Alt <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast=</span>contrast[,<span class="st">"MOV10KO_vs_CTRL"</span>], <span class="at">alpha =</span> <span class="fl">0.05</span>)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>res_tableKO_shrAlt <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds,</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">res =</span> res_tableKO_Alt,</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>                                <span class="at">type=</span><span class="st">"ashr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using 'ashr' for LFC shrinkage. If used in published research, please cite:
    Stephens, M. (2016) False discovery rates: a new deal. Biostatistics, 18:2.
    https://doi.org/10.1093/biostatistics/kxw041</code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># confirm they are the same</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>res_tableKO_Alt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): 0,+1,0 
Wald test p-value: 0,+1,0 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE      stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835       0.019599 0.0857587  0.228537 0.819228863
ENSG00000000005    26.2489      -0.561368 0.5005704 -1.121456 0.262093841
ENSG00000000419  1478.2512       0.200415 0.1267970  1.580593 0.113971138
ENSG00000000457   518.4220       0.421334 0.1113780  3.782920 0.000154999
ENSG00000000460  1159.7761       0.165758 0.0887861  1.866938 0.061910196
...                    ...            ...       ...       ...         ...
ENSG00000285889    1.82171     -4.9988725 4.4250065 -1.129687   0.2586081
ENSG00000285950    7.58089      0.1244283 1.1406775  0.109083   0.9131368
ENSG00000285976 4676.24904     -0.0641105 0.0733078 -0.874539   0.3818246
ENSG00000285978    2.25697      4.3059932 2.2301751  1.930787   0.0535094
ENSG00000285980    0.00000             NA        NA        NA          NA
                      padj
                 &lt;numeric&gt;
ENSG00000000003 0.92347051
ENSG00000000005 0.51683045
ENSG00000000419 0.31379168
ENSG00000000457 0.00243228
ENSG00000000460 0.21169526
...                    ...
ENSG00000285889         NA
ENSG00000285950         NA
ENSG00000285976   0.635781
ENSG00000285978         NA
ENSG00000285980         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>res_tableKD</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): sampletype MOV10_knockdown vs control 
Wald test p-value: sampletype MOV10 knockdown vs control 
DataFrame with 57761 rows and 6 columns
                  baseMean log2FoldChange     lfcSE      stat      pvalue
                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
ENSG00000000003  3525.8835       0.019599 0.0857587  0.228537 0.819228863
ENSG00000000005    26.2489      -0.561368 0.5005704 -1.121456 0.262093841
ENSG00000000419  1478.2512       0.200415 0.1267970  1.580593 0.113971138
ENSG00000000457   518.4220       0.421334 0.1113780  3.782920 0.000154999
ENSG00000000460  1159.7761       0.165758 0.0887861  1.866938 0.061910196
...                    ...            ...       ...       ...         ...
ENSG00000285889    1.82171     -4.9988725 4.4250065 -1.129687   0.2586081
ENSG00000285950    7.58089      0.1244283 1.1406775  0.109083   0.9131368
ENSG00000285976 4676.24904     -0.0641105 0.0733078 -0.874539   0.3818246
ENSG00000285978    2.25697      4.3059932 2.2301751  1.930787   0.0535094
ENSG00000285980    0.00000             NA        NA        NA          NA
                      padj
                 &lt;numeric&gt;
ENSG00000000003 0.92347051
ENSG00000000005 0.51683045
ENSG00000000419 0.31379168
ENSG00000000457 0.00243228
ENSG00000000460 0.21169526
...                    ...
ENSG00000285889         NA
ENSG00000285950         NA
ENSG00000285976   0.635781
ENSG00000285978         NA
ENSG00000285980         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res_tableKO_Alt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 1510, 3.9%
LFC &lt; 0 (down)     : 1300, 3.3%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res_tableKD)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 38903 with nonzero total read count
adjusted p-value &lt; 0.05
LFC &gt; 0 (up)       : 1510, 3.9%
LFC &lt; 0 (down)     : 1300, 3.3%
outliers [1]       : 28, 0.072%
low counts [2]     : 21313, 55%
(mean count &lt; 16)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>